### PR TITLE
[SPARK-35041][SQL] Revise the overflow in UTF8String

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add overflow check before do `new byte[]`.

### Why are the changes needed?

Avoid overflow in extreme case.

### Does this PR introduce _any_ user-facing change?

Maybe yes, the error msg changed if overflow.

### How was this patch tested?

Pass CI.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @maropu @MaxGekk

**Review by MaxGekk — COMMENTED** [review]
How about:
- https://github.com/apache/spark/blob/1be10124977bc32b580e37cf533328195ab6378c/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L767
- I would add an assert here https://github.com/apache/spark/blob/1be10124977bc32b580e37cf533328195ab6378c/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L886-L887 to be sure that size doesn't overflow
- Replace Ints.checkedCast at https://github.com/apache/spark/blob/1be10124977bc32b580e37cf533328195ab6378c/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L974

**Comment by ulysses-you** [comment]
> int size = numBytes - delim.numBytes - idx; 
>  byte[] bytes = new byte[size]; 

The second place seems will never overflow. Any reason to assert not overflow here ?

**Comment by MaxGekk** [comment]
> The second place seems will never overflow. Any reason to assert not overflow here ?

ok. It seems we can guarantee that `numBytes > idx + delim.numBytes` and size <= `Int.MaxValue`

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -764,7 +763,8 @@ public UTF8String repeat(int times) {
       return EMPTY_UTF8;
     }
 
-    byte[] newBytes = new byte[numBytes * times];
+    int resultSize = Math.toIntExact((long)numBytes * times);
+    byte[] newBytes = new byte[resultSize];
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:767]:
```suggestion
    byte[] newBytes = new byte[Math.multiplyExact(numBytes, times)];
```

Comment (ulysses-you) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: thanks for the carefully review !

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @ulysses-you .

**Comment by ulysses-you** [comment]
thanks for merging

**Pull request closed without merging**